lib3h_protocol/src/lib.rs

```diff
@@ -1 +1,11 @@
+//! This module provides the api definition for working with lib3h
 
+pub mod data_types;
+pub mod network_engine;
+pub mod protocol;
+
+/// Opaque Address Bytes
+pub type Address = Vec<u8>;
+
+use failure::Error;
```

Comment:
Can we use a custom `error` implementation here ? I noticed some time ago that `Failure` ,despite being super useful, suffer big performance issues, especially on a happy paths (tho I suppose from the use case later on the code right ?)

From the official Failure documentation we can read:

> The `Error` type allocates. There are cases where this would be too expensive. In those cases you should use a custom failure.

And [here is reddit thread about those perf issues] (https://www.reddit.com/r/rust/comments/7te8si/personal_experience_can_cause_hard_to_notice/)

I am happy to implement the custom error myself if it can help you :)

What do you guys think ?

Reply:
I needed to move SodiumError into lib3h_protocol to integrate with core: https://github.com/holochain/lib3h/pull/57 - let me know how that fits into you guys' thoughts

Reply:
`SodiumError into lib3h_protocol` sounds like a bad idea to me. What's the reason for that?

Reply:
If you have other ideas, totally let me know!

HolochainError needs a `From` defined for SodiumError (now Lib3hError). Previously, this was done in the sodium crate, requiring a dependency of the core_types crate. I don't think we want any crates in lib3h to depend on crates in core, and we can't make core_types depend on sodium, because core types needs to compile to WASM... hence adding it to lib3h_protocol which can be a dependency of both core_types and sodium.

Reply:
I think this needs F2F discussion. Aren't we going to create a new crate for common types?
I still believe Core should break up `core_types` into WASM and non-WASM crates.

Reply:
Following the discussion on Mattermost, I'm writing here my proposal for a replacement of `Failure` for a custom `error` handling.

I would like for performance reason using a custom error handling like the one implement here, in [rust-csv](https://github.com/BurntSushi/rust-csv/blob/master/src/error.rs).

It requires a little bit of work but it will let us have a nice and fine grained error handling.